src/Func.cpp

```diff
@@ -327,8 +339,392 @@ std::string Stage::dump_argument_list() const {
+        [&s](const ReductionVariable& rv) { return var_name_match(s.inner, rv.var, false); });
+
+    if ((iter_outer != rvars.end()) && (iter_inner != rvars.end())) {
+        debug(4) << "  Fusing " << s.outer << " and " << s.inner << " into " << s.old_var << "\n";
+        Expr extent = iter_outer->extent * iter_inner->extent;
+        iter_outer->var = s.old_var;
+        iter_outer->min = 0;
+        iter_outer->extent = extent;
+        rvars.erase(iter_inner);
+    }
+}
+
+// Apply purify directives on the reduction variables. Purify replace a RVar
+// with a Var, thus, the RVar needs to be removed from the list.
+void apply_purify(const Split &s, vector<ReductionVariable> &rvars) {
+    internal_assert(s.is_purify());
+    const auto iter = std::find_if(rvars.begin(), rvars.end(),
+        [&s](const ReductionVariable& rv) { return var_name_match(s.old_var, rv.var, false); });
+    if (iter != rvars.end()) {
+        debug(4) << "  Purify RVar " << iter->var << " into Var " << s.outer
+                 << ", deleting it from the rvars list\n";
+        rvars.erase(iter);
+    }
+}
+
+// Apply rename directives on the reduction variables.
+void apply_rename(const Split &s, vector<ReductionVariable> &rvars) {
+    internal_assert(s.is_rename());
+    const auto iter = std::find_if(rvars.begin(), rvars.end(),
+        [&s](const ReductionVariable& rv) { return var_name_match(s.old_var, rv.var, false); });
+    if (iter != rvars.end()) {
+        debug(4) << "  Renaming " << iter->var << " into " << s.outer << "\n";
+        iter->var = s.outer;
+    }
+}
+
+// Apply scheduling directives (e.g. split, fuse, etc.) on the reduction variables.
+void apply_split_directive(const Split &s, vector<ReductionVariable> &rvars) {
+    if (s.is_split()) {
+        apply_split(s, rvars);
+    } else if (s.is_fuse()) {
+        apply_fuse(s, rvars);
+    } else if (s.is_purify()) {
+        apply_purify(s, rvars);
+    } else {
+        apply_rename(s, rvars);
+    }
+}
+
+} // anonymous namespace
+
+Func Stage::rfactor(RVar r, Var v) {
+    return rfactor({{r, v}});
+}
+
+Func Stage::rfactor(vector<pair<RVar, Var>> preserved) {
+    user_assert(!definition.is_init()) << "rfactor() must be called on an update definition\n";
+
+    string func_name;
+    {
+        vector<std::string> tmp = split_string(stage_name, ".update(");
+        internal_assert(!tmp.empty() && !tmp[0].empty());
+        func_name = tmp[0];
+    }
+
+    vector<Expr> &args = definition.args();
+    vector<Expr> &values = definition.values();
+
+    // Check whether the operator is associative and determine the operator and
+    // its identity for each value in the definition if it is a Tuple
+    bool is_assoc;
+    vector<AssociativeOp> ops;
+    std::tie(is_assoc, ops) = prove_associativity(func_name, args, values);
+    user_assert(is_assoc) << "Failed to call rfactor() since it can't prove associativity of the operator\n";
```

Comment:
I think the error message needs to be more helpful. E.g. it should print out the func name which update definition it is.